Fix for arbitrary code execution while unpickling in ipycache.load_vars() method.

adi928 opened the pull request:
Have an Unpickler which would check if the deserialized bytes stream a built-in function and restrict it if neccessary. Like [here](https://github.com/moreati/pickle-fuzz#pickle)
Any malicious command trying to process through the unpickle command would have to go through the restricted_loads() method which only allows io.StringsIO to parse.
Anything else, and it would raise a UnpicklingError.

adi928 committed:
Fix for arbitrary code execution while unpickling in ipycache.load_vars() method.

Resolution for issue #47 in the original repo.

Any malicious command trying to process through the unpickle command would have to go through the restricted_loads() method which only allows io.StringsIO to parse.
Anything else, and it would raise a UnpicklingError.

adi928 committed:
Fix for arbitrary code execution while unpickling in ipycache.load_vars() method.

Resolution for issue #47 in the original repo.

Any malicious command trying to process through the unpickle command would have to go through the restricted_loads() method which only allows io.StringsIO to parse.
Anything else, and it would raise a UnpicklingError.

adi928 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	ipycache.py
#	test_ipycache.py

bropolicy commented on ipycache.py:
```diff
@@ -151,8 +151,26 @@ def save_vars(path, vars_d):
+class RestrictedUnpickler(pickle.Unpickler):
+
+    def find_class(self, module, name):
+        if module == '_io' and name == 'StringIO':
```
Can we manage to abstract the module and class name in a set?
To ease the addition of new modules in the white-list.

toufik-airane commented on ipycache.py:
```diff
@@ -151,8 +151,26 @@ def save_vars(path, vars_d):
+class RestrictedUnpickler(pickle.Unpickler):
+
+    def find_class(self, module, name):
+        if module == '_io' and name == 'StringIO':
```
`if module in safe_modules and name in safe_builtins:`

JamieSlome replied:
Closing conversation - resolution reached.

toufik-airane commented on ipycache.py:
```diff
@@ -151,8 +151,26 @@ def save_vars(path, vars_d):
+# This function provides restriction of using only the io module
+# ------------------------------------------------------------------------------
+class RestrictedUnpickler(pickle.Unpickler):
+
```
```
safe_modules = {
    '_io',
    'builtins'
}

safe_builtins = {
    'StringIO'
}
```

toufik-airane replied:
Reference: https://docs.python.org/3/library/pickle.html#restricting-globals

JamieSlome replied:
Closing conversation - resolution reached.

toufik-airane reviewed (changes requested):
Can we manage to abstract the module and class name in a set?
It will ease the addition of new modules in the whitelist to handle further needs.

JamieSlome commented:
@adi928 - thank you for your submission! Our sheriff @bropolicy has made a comment on your code. If you are able to meet the requested change within the next 48 hours, the bounty is yours. Otherwise, we will push it back to the community - good luck.

adi928 committed:
Addressing review comments for:
1. ipycache.py - Abstracting the module and class name in a set
2. test_ipycache.py - Commenting and removal of the test file.

adi928 commented:
Addressed the review comments. Couple of things I noticed:

1. I can move the IPython.utils.traitlets to a top-level since that is giving a warning as of now. 
2. It seems to me that 'cPickle' in python 2.7 is also referred to as 'pickle', so we can move the import statements out of the python version checks.

I haven't included these changes since they are not directly related to the issue being addressed by this pull request, but I can go ahead and include them.

toufik-airane reviewed (approved):
LGTM. 👏
Thank you for your contribution.

JamieSlome commented:
Congratulations @adi928 - you are the successful huntr!
Thank you for submitting your fix - we have approved your pull request and will merge your commits into our master branch. 
If you have any questions, please respond in the comments section. Your bounty is on its way - keep hunting.
